pinot-common/src/main/java/org/apache/pinot/common/utils/FileUploadDownloadClient.java

```diff
@@ -727,6 +727,23 @@ public SimpleHttpResponse uploadSegment(URI uri, String segmentName, InputStream
     return uploadSegment(uri, segmentName, inputStream, null, parameters, DEFAULT_SOCKET_TIMEOUT_MS);
   }
 
+  /**
+   * Controller periodic task uses this endpoint to ask servers to upload committed llc segment to segment store if missing.
+   * @param uri The uri to ask servers to upload segment to segment store
+   * @return the uploaded segment download url from segment store
+   * @throws URISyntaxException
+   * @throws IOException
+   * @throws HttpErrorStatusException
+   */
+  public String uploadToSegmentStore(String uri)
```

Comment:
I wouldn't put this method in this class. This class is used to upload/download files from controller. While this method is sending request to the servers. You can just add a helper method into the `PinotLLCRealtimeSegmentManager`.

Reply:
+1, I agree with Jackie.

Reply:
How was this resolved?

Reply:
This function was moved back to this class after we realized we need to utilize  `SSLContext` in `FileUploadDownloadClient`.

Reply:
I also thought about this. But I don't want to make this endpoint as a generic post, because it doesn't actually post with any payload.

Reply:
@liuchang0520 This class is responsible for uploading/downloading files. Adding this new method to this class extends the responsibility of this class - which is against the "Single Responsibility Principle".
I suggest we create a new class, `DeepStoreMissingSegmentCorrector` (or maybe some better name?). This new class has one public method `uploadToDeepStoreIfMissing`. Instead of adding `uploadToSegmentStore` to FileUploadDownloadClient, we can create SSLContext and httpClient in the constructor of the new class and use the http client in `uploadToDeepStoreIfMissing` method to instruct servers to upload missing segments to deep store.
The instance for this new class can be instantiated in `BaseControllerStarter` right after instantiation of `_pinotLLCRealtimeSegmentManager`. You can pass in sslConfig subset required for creating SSLContext object as the constructor parameter. Later on in baseControllerStarter, this instance can be injected to `RealtimeSegmentValidationManager` object for triggering missing segment upload to deep store.
With this refactoring, all the new logic will be in one place and it also doesn't add new stuff to PinotLLCRealtimeManager which is already big.
We can chat on slack or zoom if I didn't do a good job explaining the refactoring.

Reply:
Hi @sajjad-moradi , thanks for detailed instruction. I added a TODO for this. Will address this in the future. Currently we need to merge this change ASAP as we are facing a prod issue.